src/nyx/gpu/geomoments_raw.cu

```diff
@@ -0,0 +1,224 @@
+#include <cub/cub.cuh>
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <cuda_runtime_api.h>
+#include <builtin_types.h>
+#include "gpu.h"
+#include "../features/pixel.h"
+
+namespace Nyxus
+{
+    extern double* devPrereduce;    // reduction helper [roi_cloud_len]
+    extern double* d_out;           // 1 double
+    extern void* d_temp_storage;    // allocated [] elements by cub::DeviceReduce::Sum()
+    extern size_t temp_storage_bytes;
+};
+
+__device__ double pow_pos_int_raw (double a, int b)
+{
+    double retval = 1.0;
+    for (int i = 0; i < b; i++)
+        retval *= a;
+    return retval;
+}
+
+__global__ void kerRawMoment (
+    double* d_prereduce,
+    const Pixel2* d_roicloud,
+    size_t cloudlen,
+    StatsInt base_x, 
+    StatsInt base_y,
+    int p,
+    int q)
+{
+    int tid = threadIdx.x + blockIdx.x * blockSize;
+
+    if (tid >= cloudlen)
+        return;
+
+    float inten_ = d_roicloud[tid].inten,
+        x_ = d_roicloud[tid].x,
+        y_ = d_roicloud[tid].y, 
+        x0 = base_x, 
+        y0 = base_y,
+        localX = x_ - x0,
+        localY = y_ - y0;
+    
+    d_prereduce[tid] = inten_
+        * pow_pos_int_raw(localX, p)
+        * pow_pos_int_raw(localY, q);
+}
+
+__global__ void kerRawMomentWeighted (
+    double* d_prereduce,
+    const RealPixIntens* d_realintens, 
+    const Pixel2* d_roicloud,
+    size_t cloudlen,
+    StatsInt base_x,
+    StatsInt base_y,
+    int p,
+    int q)
+{
+    int tid = threadIdx.x + blockIdx.x * blockSize;
+
+    if (tid >= cloudlen)
+        return;
+
+    d_prereduce[tid] = d_realintens[tid]
+        * pow_pos_int_raw(double(d_roicloud[tid].x - base_x), p) 
+        * pow_pos_int_raw(double(d_roicloud[tid].y - base_y), q);
+}
+
+bool drvRawMoment (
+    double &retval, 
+    int p, int q,
+    const Pixel2 *d_roicloud,
+    size_t cloudlen,
+    StatsInt base_x,
+    StatsInt base_y)
+{
+    int nblo = whole_chunks2(cloudlen, blockSize);
+    kerRawMoment <<< nblo, blockSize >>> (Nyxus::devPrereduce, d_roicloud, cloudlen, base_x, base_y, p, q);
+
+    CHECKERR (cudaPeekAtLastError());
+    CHECKERR (cudaGetLastError());   

```

Comment:
Do we need both cudaPeekAtLastError and cudaGetLastError?

Reply:
Having cudaPeekAtLastError() is the maximum defensive way to run kernels because unlike the pair cudaDeviceSynchronize()-cudaGetLastError() checking the error status of all the finished threads, the former checks the thread blocks launch phase, for example, if a block of requested size is available (someone may tweak #define blockSize in the header) or if requested amount of shared memory can be allocated. I'd leave cudaPeekAtLastError() until we fully optimize and freeze the launch parameters (the block size and kernel).